Fixes #31665 - Remove changeTaxonomy actions from layout

tbrisker opened the pull request:
Instead of passing the current taxonomies all the way down from the app,
use the ForemanContext hooks to get the current one inside the
taxonomies switcher.


<!---

Thank you for contributing to The Foreman, please read the
[following guide](https://www.theforeman.org/contribute.html), in short:

* [Create an issue](https://projects.theforeman.org/projects/foreman/issues)
* Reference the issue via `Fixes #1234` in the commit message
* Prefer present-tense, imperative-style commit messages
* Mark all strings for translation, see [1]
* Suggest prerequisites for testing and testing scenarios following example above.
* Prepend `[WIP]` for work in progress to prevent bots from triggering actions
* Be patient, we will do our best to take a look as soon as we can
* Explain the purpose of the PR, attach screenshots if applicable
* List all prerequisites for testing (e.g. VMware cluster, two smart proxies...)
* Reviewers often use extensive list of items to check, have a look prior submitting [2]
* Be nice and respectful

1: https://projects.theforeman.org/projects/foreman/wiki/Translating#Translating-for-developers
2: https://github.com/theforeman/foreman/blob/develop/developer_docs/pr_review.asciidoc
-->

tbrisker commented:
@amirfefer can you take a look please? I'm having a hard time getting the context hooks to work properly inside tests, I tried several different ways of handling it but no luck so far.

tbrisker commented:
i _think_ i got the tests fixed, but i'm not sure it's the correct fix.

tbrisker commented:
Tests should be good now, ready for review

MariaAga commented:
@tbrisker The switcher should be able to change the current taxonomy value, but from what I understand if we're using `getForemanContext` there is no way to set it's value. Also if we're using this method we can probably drop `LAYOUT_CHANGE_LOCATION` and `LAYOUT_CHANGE_ORG`

tbrisker commented:
@MariaAga currently, changing taxonomy triggers a full page reload so the context is populated with the new taxonomy, meaning the actions aren't really being used at all. Theoretically we could also set the context but i'm not sure it really makes sense as we might end up in a situation where the client and server aren't in agreement on what is the current taxonomy. I'll drop the actions as well since they don't really make sense.

tbrisker committed:
Fixes #31665 - Remove changeTaxonomy actions from layout

Changing a taxonomy triggers a full page reload, so `changeLocation` and
`changeOrganization` have no actual use. Dropping them also simplifies
components quite a bit.

tbrisker commented:
Added a separate commit to drop the actions and all code related to them.

tbrisker commented on webpack/assets/javascripts/react_app/components/Layout/LayoutHelper.js:
```diff
@@ -80,9 +75,8 @@ const createOrgItem = orgs => {
   orgs.forEach(org => {
     const childObject = {
       type: org.type,
-      name: isEmpty(org.title) ? org.title : __(org.title),
```
not directly related, but this is a mistake, taxonomy titles shouldn't be translated

ezr-ondrej commented:
[test foreman] failures doesn't seem relevant and quite random only on Ruby 2.6, but lets see

ezr-ondrej reviewed (approved):
Loving this :heart: everything works without change with `-500` lines :tada:

MariaAga commented:
@tbrisker The actions are there for 2 reasons:
1. the full page reload takes a long time and the user needs an indication that something has change, as a middle way solution I changed the actions in my PR (https://github.com/theforeman/foreman/pull/7805) to show loading when the action is called.
2. We want to be able to eventually maybe not use full page reload for changing taxonomies.  

I think @sharvit had some comments on this change he wanted to raise as well.

ezr-ondrej commented:
> 1. the user needs an indication that something has change,

I don't agree with point 1. that is very SPA mindset. The user has clear indication - his browser is telling him the page is loading the old school style (before ajax fancy loaders came in).

> 2. We want to be able to eventually maybe not use full page reload for changing taxonomies.

I agree with this point, but it will yet take a while for that to happen, and it is not that hard to add these actions, once we will get there :)

tbrisker commented:
To add to point 2 - since we changed the way we get taxonomies to using the context instead of redux, the action would have to be written completely differently even if we do eventually move it to a more SPA approach.

sharvit commented:
I am not sure why need to migrate from redux to react-context.
If the goal is not to "passing the current taxonomies all the way down from the app", a solution would be to create a hook that brings data from redux to the component that needs it.
```js
const useForemanLocation = () => useSelector(selectForemanLocation);
const useForemanOrganization = () => useSelector(selectForemanOrganization);
```

That way we can continue using our current actions and selectors and achieve better performance and a nicer interface.

My main concern is about having too many sources of truth which leads to uncertainty and cupelling of data.

The original goal of the foreman-context was to provide easy access to server metadata that doesn't change during the UI lifecycle (app-version, app-settings, app-docs) and that's the reason we never introduced actions to change to the context.

I feel that a safer approach would be to continue using redux for the global-state and react-context for global-metadata.

tbrisker commented:
@sharvit https://github.com/theforeman/foreman/pull/7655 added the taxonomies to the context, since they are useful in many components and rarely change, and the intent was to use the context as the single source of truth for them. https://github.com/theforeman/foreman/pull/8267 deprecated the use of the store as the source for the current taxonomies. This PR starts to clean up the current places that use the store instead of the context. There are a couple of other places I plan to refactor following this PR but did not want to make it too large and remove it completely.
I believe this PR shows the main motivation for using the context instead of the store in this case - greatly simplified components that are easier to maintain. As mentioned already in the PR adding the taxonomies to the context, if in the future we want to update the context from the front end, that can also be done and will be much simpler than the way it is currently being done. Even if we want to show a spinner while the page reloads with new taxonomies that doesn't require storing them in the store.

ezr-ondrej reviewed (dismissed):
This is in line with what we say in [storybook](https://github.com/theforeman/foreman/blob/26a20e665410b512a8dbba4287a7944330a2535d/webpack/stories/docs/state-management.stories.mdx#foremancontext) and deprecation of the selector. If someone strongly disagree please open a discourse post and change the course we are taking to the ForemanContext, until we do, this is going the direction we've set..

tbrisker commented:
Updating this PR on top of @MariaAga's changes. Her PR already included all of the first commit in the original PR which changed the switchers to use the context, the remainder in this PR is removing the taxonomy change actions from the store and components using them.

ezr-ondrej reviewed (approved):
Nice cleanup :+1:
Thanks @tbrisker !
